Civi/Api4/Action/CaseSalesOrder/ContributionCreateAction.php

```diff
@@ -0,0 +1,186 @@
+<?php
+
+namespace Civi\Api4\Action\CaseSalesOrder;
+
+use Civi\Api4\CaseSalesOrder;
+use Civi\Api4\PriceFieldValue;
+use Civi\Api4\PriceField;
+use Civi\Api4\PriceSet;
+use CRM_Core_Transaction;
+use Civi\Api4\Generic\Result;
+use Civi\Api4\Generic\AbstractAction;
+use Civi\Api4\Generic\Traits\DAOActionTrait;
+use CRM_Contribute_BAO_Contribution as Contribution;
+use CRM_Civicase_Service_CaseSalesOrderLineItemsGenerator as salesOrderlineItemGenerator;
+use Civi\Api4\CaseSalesOrderContribution as Api4CaseSalesOrderContribution;
+
+/**
+ * Creates contribution for multiple sales orders.
+ */
+class ContributionCreateAction extends AbstractAction {
+  use DAOActionTrait;
+
+  /**
+   * Sales order IDs.
+   *
+   * @var array
+   */
+  protected $salesOrderIds;
+
+  /**
+   * Sales order Status ID.
+   *
+   * @var int
+   */
+  protected $statusId;
+
+  /**
+   * Type of invoice (either percent or remain).
+   *
+   * @var string
+   */
+  protected $toBeInvoiced;
+
+  /**
+   * The percentage value.
+   *
+   * @var int
+   */
+  protected $percentValue;
+
+  /**
+   * Contribution Date.
+   *
+   * @var string
+   */
+  protected $date;
+
+  /**
+   * Contribution Financial Type ID.
+   *
+   * @var string
+   */
+  protected $financialTypeId;
+
+  /**
+   * {@inheritDoc}
+   */
+  public function _run(Result $result) { // phpcs:ignore
+    $resultArray = $this->createContribution();
+
+    $result->exchangeArray($resultArray);
+  }
+
+  /**
+   * {@inheritDoc}
+   */
+  protected function createContribution() {
+    $transaction = CRM_Core_Transaction::create();
+    $priceField = $this->getDefaultContributionPriceField();
```

Comment:
you here fetch all the default price fields and not just one, so it might be better to rename this to:

getDefaultPriceSetFields

and $priceField to $defaultPriceFields